### DISCUSSION TITLE
Need some guidance opening RAW RGB data

### DISCUSSION THREAD

**Discussion by Jeroen66** [dlemstra/Magick.NET#1237] [Help]
Ok, been playing with this for a few days now, but don't seem able to find the correct magical combination of defines to this type of file.

It is a file from a Pakon scanner, 3000x2000 pixels, 16-bit, three channels with a 16 byte header. With those values Photoshop can open it, so these must be right. 

In ImageMagick that would be a size of 3000x2000+16, but that does not seem to go well with Magick.NET, so I believe I need a way to define that 16 byte header/offset somewhere else. Or maybe I have made another mistake. I'm on Visual Basic BTW.

Tried a lot of things, now my code looks like this:

 Dim bytes = My.Computer.FileSystem.ReadAllBytes(RawfileName)
        Dim RAWDefines As New MagickReadSettings
        RAWDefines.SetDefine(MagickFormat.Rgb, "size", "2000x3000")
        RAWDefines.SetDefine(MagickFormat.Rgb, "depth", "16")
        RAWDefines.SetDefine(MagickFormat.Rgb, "interlace", "plane")
        'and we should add something here as the file has a 16 byte offset
        Dim RawImage As New MagickImage(bytes, RAWDefines)
        PictureBox1.Image = RawImage.ToBitmap

This results in the error "no decode delegate for this image format `' @ error/blob.c/BlobToImage/460". Maybe because of the .RAW extension?

Any insight would be greatly appreciated. 

Thanks
Jeroen

**Comment by dlemstra** [comment]
We don't have something build in to ImageMagick to read an image like this. But it might be something we could add in the future if you happen to have a reference document of the format. For now I would advise you to read the image like this. The code below is not tested but I hope it might give you a few ideas how you can solve this:

```C#
byte[] bytes;
var lengthInBytes = 2000 * 3000 * 2;
var offset = 16;

var settings = new PixelReadSettings(2000, 3000, StorageType.Short, "r");
using var image = new MagickImage();
image.ReadPixels(bytes, offset, lengthInBytes, settings);

settings.Mapping = "g";
using var green = new MagickImage();
green.ReadPixels(bytes, (offset * 2) + lengthInBytes, lengthInBytes, settings);
image.Composite(green, Channels.Green);

settings.Mapping = "b";
using var blue = new MagickImage();
blue.ReadPixels(bytes, (offset * 3) + (2 * lengthInBytes), lengthInBytes, settings);
image.Composite(blue, Channels.Blue);
```

I might also have an other idea to make this easier. Would it be possible to share an example file? Feel free to reach out by email if you don't want to share that file publicly.

**Reply by Jeroen66** [reply]
Hi Dirk,

Thank you for looking into this. (and for magick.net of course!)

I will wetransfer an example. If I succeed doing this in VB.net I will post my code here as well.

Thanks

Jeroen

**Reply by dlemstra** [reply]
I have added a method to make it easier to import raw pixel data. The above example can now also be written like this:

```C#
byte[] bytes;
var lengthInBytes = image.Width * image.Height * 2;
var offset = 16;

using var image = new MagickImage(MagickColors.Red, 3000, 2000);
var settings = new PixelImportSettings(image.Width, image.Height, StorageType.Short, "r");
image.ImportPixels(bytes, offset, settings);
offset += lengthInBytes;
settings.Mapping = "g";
image.ImportPixels(bytes, offset, settings);
offset += lengthInBytes;
settings.Mapping = "b";
image.ImportPixels(bytes, offset, settings);
```
